File: branches/split_tools_and_core/mccode-legacy-tools/support/common/pgplot/sys_linux/f95_src/grgenv.f
C*GRGENV -- get value of PGPLOT environment parameter (NAGWare f95)
C+
      SUBROUTINE GRGENV(NAME, VALUE, L)
      USE F90_UNIX_ENV, ONLY : GETENV
      CHARACTER*(*) NAME, VALUE
      INTEGER L
C
C Return the value of a PGPLOT environment parameter. In Sun/Convex-UNIX,
C environment parameters are UNIX environment variables; e.g. parameter
C ENVOPT is environment variable PGPLOT_ENVOPT. Translation is not
C recursive and is case-sensitive.
C
C Arguments:
C  NAME   : (input) the name of the parameter to evaluate.
C  VALUE  : receives the value of the parameter, truncated or extended
C           with blanks as necessary. If the parameter is undefined,
C           a blank string is returned.
C  L      : receives the number of characters in VALUE, excluding
C           trailing blanks. If the parameter is undefined, zero is
C           returned.
C--
C 25-Mar-1988
C-----------------------------------------------------------------------
      INTEGER I, LIN, ERRNO
      CHARACTER*32 TEST
C
      TEST = 'PGPLOT_'//NAME
      LIN = INDEX(TEST, ' ')-1
      ERRNO = 0
      CALL GETENV(TEST(:LIN), VALUE, ERRNO=ERRNO)
      IF (ERRNO.NE.0 .OR. VALUE.EQ.' ') THEN
          L = 0
      ELSE
          DO 10 I=LEN(VALUE),1,-1
              L = I
              IF (VALUE(I:I).NE.' ') GOTO 20
   10     CONTINUE
          L = 0
   20     CONTINUE
      END IF
      END

File: branches/split_tools_and_core/mccode-legacy-tools/support/common/pgplot/sys_cray/grgenv.c
#include <fortran.h>
#include <stdlib.h>
#include <string.h>

/*
 **GRGENV -- get value of PGPLOT environment parameter (Cray)
 *+
 *    SUBROUTINE GRGENV(NAME, VALUE, L)
 *    CHARACTER*(*) NAME, VALUE
 *    INTEGER L
 *
 * Return the value of a PGPLOT environment parameter. In Sun/Convex-UNIX,
 * environment parameters are UNIX environment variables; e.g. parameter
 * ENVOPT is environment variable PGPLOT_ENVOPT. Translation is not
 * recursive and is case-sensitive.
 *
 * Arguments:
 *  NAME   : (input) the name of the parameter to evaluate.
 *  VALUE  : receives the value of the parameter, truncated or extended
 *           with blanks as necessary. If the parameter is undefined,
 *           a blank string is returned.
 *  L      : receives the number of characters in VALUE, excluding
 *           trailing blanks. If the parameter is undefined, zero is
 *           returned.
 *--
 * 09-Nov-1994 - [mcs] Fortran callable C version for CRAY.
 *-----------------------------------------------------------------------
 */
void GRGENV(_fcd fortran_name, _fcd fortran_value, int *length)
{
  static char *prefix = "PGPLOT_";  /* Environment variable name prefix */
  char test[33];     /* PGPLOT_* Concatenation buffer */
  int name_len;      /* Un-padded length of 'name' string */
  int prefix_len;    /* The length of prefix[] */
  char *env=0;       /* Environment variable value */
  int i;
/*
 * Extract the string pointer and dimension of each FORTRAN string argument.
 */
  char *name = _fcdtocp(fortran_name);
  int name_dim = _fcdlen(fortran_name);
  char *value = _fcdtocp(fortran_value);
  int value_dim = _fcdlen(fortran_value);
/*
 * Determine the length of 'name' by searching for the last
 * non-space character.
 */
  name_len = name_dim;
  while(name_len > 0 && name[name_len-1] == ' ')
    name_len--;
/*
 * Determine the length of the prefix.
 */
  prefix_len = strlen(prefix);
/*
 * Prefix 'name' with PGPLOT_ if there is room in test[].
 */
  if(prefix_len + name_len + 1 <= sizeof(test)/sizeof(char)) {
    strcpy(test, prefix);
    strncpy(&test[prefix_len], name, name_len);
    test[prefix_len+name_len] = '\0';
/*
 * Get the value of the environment variable now named in test[].
 */
    env = getenv(test);
  };
/*
 * Substitute an empty string if no value was obtained, or the value
 * obtained is too long to fit in the output string.
 */
  if(env==0 || strlen(env) > value_dim)
    env = "";
/*
 * Copy the environment variable value into the output string.
 */
  strncpy(value, env, value_dim);
/*
 * Return the unpadded length of the string.
 */
  {
    int env_len = strlen(env);
    *length = (env_len <= value_dim) ? env_len : value_dim;
  };
/*
 * Pad the fortran string with spaces.
 */
  for(i = *length; i<value_dim; i++)
    value[i] = ' ';
  return;
}
